snapshotter/demux/metrics/proxy.go

```diff
@@ -30,10 +32,71 @@ const (
 	MaxMetricsResponseSize = 32768 // 32 KB
 )
 
+// Monitor is used for port and lifecycle management of metrics proxies.
+type Monitor struct {
+	// monitorChan indicates a port is free for metrics proxy server.
+	ch chan int
+	// portMap tracks used and unused metrics proxy server ports.
+	portMap map[int]bool
+	// startPort is first port in the configurable range of ports to serve metrics proxies from.
+	startPort int
+	// endPort is last port in the configurable range of ports to serve metrics proxies from.
+	endPort int
+	// mu is the lock used for threadsafe reading and writing of the metrics proxy port map.
+	mu *sync.Mutex
+}
+
+// NewMonitor returns a new metrics proxy monitor.
+func NewMonitor(startPort, endPort int) *Monitor {
+	return &Monitor{make(chan int), make(map[int]bool), startPort, endPort, &sync.Mutex{}}
+}
+
+// Start receives messages from the proxy server channel indiciating a server has closed or terminated.
+// It marks the metrics proxy server's port as available.
+func (m *Monitor) Start() error {
+	for freePort := range m.ch {
+		m.mu.Lock()
+		delete(m.portMap, freePort)
```

Comment:
So once a port is free'd it is no longer available for metrics proxy allocation?

Reply:
No, the port map is really treated as a set. so the "set" is checked in `findOpenPort` to see if the element exists in it. If not, then that means the element (port) is free.